[MRG+2] Deprecate reorder parameter in auc

qinhanmin2014 opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#Contributing-Pull-Requests
-->
#### Reference Issue
<!-- Example: Fixes #1234 -->
proposed in #9786 by @lesteve 

> with this change reorder=False(reorder=True?) in the auc function is not used in our code. I think we should deprecate the reorder=False(reorder=True?) parameter and potentially (up for debate) have some threshold in case there are some small negative values in np.diff(tpr) or np.diff(fpr)

#### What does this implement/fix? Explain your changes.
#### (1)add tolerance to roc_auc_score when checking whether x is increasing. (discarded)
<!--
Currently, when setting reordering=False, auc will check whether x is increasing and raise error if not. But sometimes python will introduce slight error when doing float calculation and users will unexpectedly get an error indicating that x is not sorted. So it might be better to allow some tolerance when checking whether x is increasing.
-->
#### (2)improve error message when x is neither increasing nor decreasing.
#### (3)deprecate reorder parameter. reorder=False is now the default behaviour.
reorder=True is introduced in https://github.com/scikit-learn/scikit-learn/commit/9f18586ebfc835810afff50fcc3449265501bbcb for roc_auc_score. Now it is not used in roc_auc_score due to #9786.

#### Any other comments?
cc @jnothman @lesteve 

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

qinhanmin2014 committed:
add tol parameter

lesteve commented:
I would deprecate `reorder` in the same PR.

To be honest I am not sure it is worth adding `tol`. If the user has non increasing fpr (which means he is already in a special case, i.e. not using roc_curve which guarantees that fpr and tpr are increasing) then it is up to him to preprocess fpr and tpr before passing them to `auc`. The error message could be a bit improved to help the user do the right thing. e.g. by mentioning how many negative `np.diff(fpr)` there is and what's the most negative values we found.

qinhanmin2014 committed:
address comment

qinhanmin2014 committed:
Merge remote-tracking branch 'upstream/master' into auc_tol

qinhanmin2014 commented:
@lesteve Thanks for the instant reply :)
I have addressed all your comments. Now the PR:
(1)add tolerance to roc_auc_score when checking whether x is increasing.
(2)improve error message when x is neither increasing nor decreasing.
(3)deprecate reorder parameter. reorder=False is now the default behaviour.
If you are sure that we don't need some part of the PR, please give a clear instruction and I'll remove it.

lesteve commented:
I would be in favour of removing `tol` and making the error message more informative.

qinhanmin2014 committed:
remove tol

qinhanmin2014 committed:
error message

qinhanmin2014 commented:
@lesteve Thanks. I have removed parameter tol.
For the erorr message, I have sligtly improved it previously. Now it looks like this:
```python
auc([2, 1, 3, 4], [4, 5, 6, 7])
# x is neither increasing nor decreasing : [2 1 3 4].
# np.diff(x) contains 2 positive values and 1 negative values.
```
Since x can be either increasing or decreasing, I'm wondering what's the best information to provide.
WDYT? Thanks a lot :)

lesteve commented:
Maybe the most negative value and the most positive value of np.diff? This way you can see that his how far your array is from being increasing or decreasing.

qinhanmin2014 committed:
update error message

qinhanmin2014 commented:
@lesteve Thanks for the instant reply :) I have updated the error message.
```python
auc([2.5, 1.6, 3.7, 4.8], [5, 6, 7, 8])
# ValueError: x is neither increasing nor decreasing : [ 2.5  1.6  3.7  4.8].
# np.diff(x) contains 2 positive values and 1 negative values.
# The most positive value in np.diff(x) : x[2] - x[1] = 2.1.
# The most negative value in np.diff(x) : x[1] - x[0] = -0.9.
```

qinhanmin2014 committed:
minor fix

qinhanmin2014 committed:
minor fix

qinhanmin2014 committed:
use int in the test to avoid strange failure

lesteve commented on sklearn/metrics/tests/test_ranking.py:
```diff
@@ -426,7 +427,20 @@ def test_auc_errors():
     # x is not in order
-    assert_raises(ValueError, auc, [1.0, 0.0, 0.5], [0.0, 0.0, 0.0])
+    error_message = ("x is neither increasing nor decreasing : "
+                     "[ 2.5  1.6  3.7  4.8]. np.diff(x) contains 2 positive "
```
numpy 1.14 will change np array repr/str so don't hard-code it like this instead use `'{}'.format(your_array)`.

lesteve commented on sklearn/metrics/ranking.py:
```diff
@@ -95,8 +105,16 @@ def auc(x, y, reorder=False):
+                                 "x[{}] = {}. The most negative value in "
+                                 "np.diff(x) : x[{}] - x[{}] = {}."
+                                 .format(x, (dx > 0).sum(), (dx < 0).sum(),
+                                         dx.argmax() + 1, dx.argmax(),
```
Define the variables that you reuse outside of the .format

qinhanmin2014 committed:
improve

qinhanmin2014 commented:
@lesteve Thanks :) Both comments are addressed.

jnothman commented on sklearn/metrics/ranking.py:
```diff
@@ -47,13 +47,18 @@ def auc(x, y, reorder=False):
     y : array, shape = [n]
         y coordinates.
-    reorder : boolean, optional (default=False)
+    reorder : boolean, optional (default=None)
```
can we make the default something semantically meaningful, like 'warn' or 'deprecated', please?

jnothman commented on sklearn/metrics/ranking.py:
```diff
@@ -47,13 +47,18 @@ def auc(x, y, reorder=False):
     ----------
     x : array, shape = [n]
-        x coordinates.
+        Increasing or decreasing x coordinates.
```
How about "x coordinates. These must be either monotonic increasing or monotonic decreasing."? Otherwise, this seems vague.

jnothman commented on sklearn/metrics/ranking.py:
```diff
@@ -95,8 +105,17 @@ def auc(x, y, reorder=False):
-                                 "the x array is not increasing: %s" % x)
+                maxpos = dx.argmax()
+                minpos = dx.argmin()
+                raise ValueError("x is neither increasing nor decreasing "
```
I think this message is a bit obscure to a reader of the code, more so than to someone who receives the error by surprise. It deserves a comment, perhaps a reference to the issue.

jnothman reviewed (commented):
Adjust the title to remove reference to tol?

qinhanmin2014 committed:
improve

qinhanmin2014 committed:
minor fix

jnothman commented on sklearn/metrics/ranking.py:
```diff
@@ -95,8 +106,16 @@ def auc(x, y, reorder=False):
-                raise ValueError("Reordering is not turned on, and "
-                                 "the x array is not increasing: %s" % x)
+                # Output the most positive value and most negative value in
+                # np.diff(x) for debugging.
```
Is it meant to handle the case that adjacent values are out of order but only within a small tolerance?

jnothman commented on sklearn/metrics/ranking.py:
```diff
@@ -95,8 +106,16 @@ def auc(x, y, reorder=False):
+                minpos = dx.argmin()
+                raise ValueError("x is neither increasing nor decreasing "
+                                 ": {}. The most positive value in np.diff(x)"
+                                 " : x[{}] - x[{}] = {}. The most negative "
```
I suspect that this error message will most often leave the user bewildered.

qinhanmin2014 commented:
@jnothman Thanks :) The error message is suggested by @lesteve [here](https://github.com/scikit-learn/scikit-learn/pull/9851#issuecomment-333035322). I don't have strong opinion on this. Seems that it might indeed introduce some confusion. But at least the original error message is not right. So do you think `"x is neither increasing nor decreasing : {}".format(x)` (very similar to the original message) is enough?

jnothman commented:
Why do we want to allow increasing or decreasing? Why don't we require it to be increasing, and state that there is a negative diff of ...? I think it would be much clearer to the user why this is relevant.

jnothman commented:
It's not like we cause great harm to the user by asking them to reverse their inputs.

qinhanmin2014 commented:
@jnothman Thanks. From my perspective, I agree with you. Allow x to be decreasing seems awkward but the implementation is there. So what's your final decision or do you want me to further investigate the origin of the code?

jnothman commented on sklearn/metrics/ranking.py:
```diff
@@ -47,13 +47,19 @@ def auc(x, y, reorder=False):
+        x coordinates. These must be either monotonic increasing or monotonic
+        decreasing.
     y : array, shape = [n]
         y coordinates.
```
don't these also need to be correspondingly increasing/decreasing in case of ties?

jnothman commented on sklearn/metrics/ranking.py:
```diff
@@ -47,13 +47,19 @@ def auc(x, y, reorder=False):
         y coordinates.
-    reorder : boolean, optional (default=False)
+    reorder : boolean, optional (default='deprecated')
         If True, assume that the curve is ascending in the case of ties, as for
```
I assume this should say "if False"..???

jnothman reviewed (commented):
So the support for decreasing data seems to be undocumented? (Worth dropping support for that too? Or will it actually affect users?)

qinhanmin2014 committed:
update comment

qinhanmin2014 commented:
@jnothman Thanks for the review :)

> So the support for decreasing data seems to be undocumented?

I think the doc does not clarify what situation we support. Neither the increasing case nor the decreasing case.

> don't these also need to be correspondingly increasing/decreasing in case of ties?

Currently, when reorder is False, we don't check y and users get different result with different y order. When reorder is True, y is used to break ties for x when sorting. It seems reasonable to me.
```python
auc([1, 2, 2], [2, 3, 4])  # 2.5
auc([1, 2, 2], [2, 4, 4])  # 3.0
auc([1, 2, 2], [2, 3, 4], reorder=True)  # 2.5
auc([1, 2, 2], [2, 4, 4], reorder=True)  # 2.5
```

> I assume this should say if False

At a glance, it seems wrong. But I think the original author wants to tell us that if x is the same (in the case of ties), we need to ensure that y is increasing (the curve is ascending). Anyway, I think it is not friendly to users.

After consideration, my suggestions are
(1)Modify the document and the error message to match current implemetation.
(2)Gradually deprecate parameter reorder(it is not used in any example)
(3)Seems that the support for decreasing x is introduced in https://github.com/scikit-learn/scikit-learn/commit/9f18586ebfc835810afff50fcc3449265501bbcb by GaelVaroquaux in 2012. It has long been here and we will not need much effort to support the function. So you may need to carefully consider whether to drop the function.

jnothman commented:
fine, let's keep support for decreasing, but then let's not give a long
error message that perplexes the user. If they are surprised to find the
data is not ordered, they will have not been using roc_auc_score, and are I
suppose capable of diagnosing or sorting themselves.

And we can check if y is sorted with something like assert
all(np.where(diffx == 0, diffy, diffx) >= 0)

qinhanmin2014 committed:
reduce message

qinhanmin2014 commented:
@jnothman Thanks :)

> but then let's not give a long error message that perplexes the user. If they are surprised to find the data is not ordered, they will have not been using roc_auc_score, and are I suppose capable of diagnosing or sorting themselves.

I agree. The error message now is very similar to the previous one.

> And we can check if y is sorted with something like assert all(np.where(diffx == 0, diffy, diffx) >= 0)

Are you sure we want to do so or maybe I misunderstand you? From my perspective, if users provide something like `auc([1, 2, 2, 3], [1, 3, 2, 1])`, I think it might be better to output 3.5 (current behaviour for reorder=False) instead of raising an error.

jnothman commented:
Hmm. Yes, I suppose so.

qinhanmin2014 commented:
@jnothman Sorry but I don't understand your previous comment. What else do I need to do here? Thanks :)

jnothman commented on sklearn/metrics/ranking.py:
```diff
@@ -47,12 +47,19 @@ def auc(x, y, reorder=False):
+        used to break ties when sorting x.
+
+        ..deprecated:: 0.20
+          Parameter ``reorder`` has been deprecated in version 0.20 and will
```
please add a brief note on why

jnothman reviewed (commented):
otherwise lgtm

qinhanmin2014 committed:
add note and update what's new

qinhanmin2014 commented:
@jnothman Thanks :) Note added in doc and what's new.

jnothman commented:
Could you please remind me how reordering related to the numerical instability in `roc_auc_score`?

qinhanmin2014 commented:
@jnothman Sure :) The main reasons for the numerical instability in `roc_auc_score` are as follows:
(1) We cannot ensure that fps/fpr is increasing since we use subtraction. Actually fps/fpr is increasing not considering the slight error introduced when doing float calculation.
(2) We set reorder=True and sort the points according to x (fpr). So we sort the points into wrong order and get wrong result from roc_auc_score.
Among the two reasons, the second one is more decisive. Because even with the original implementation, if we set reorder=False, we can get result nearly equal to the right result (not consider the error raised by sanity check in auc).
In #9786, we actually solve the problem in a better way : ensure that fps/fpr is increasing by using accumulation and set reorder=False. In this way, we pass the sanity check in auc and get more precise result.
I illustrate the problem with a plot [here](https://github.com/scikit-learn/scikit-learn/pull/9786#discussion_r139913765). You may have a look if you want to further investigate the problem.

jnothman commented:
Yes, okay. So to be sure, there's nothing wrong with providing the user
with an option to reorder...

I'm not entirely convinced of the merits of this PR then. Deprecation and
removal without clear cause breaks someone's code. Why do we feel it is
appropriate/necessary to remove an existing feature (albeit of no need in
the library)?

qinhanmin2014 commented:
@jnothman Personally, I would prefer to remain reorder parameter. Like the support for decreasing x, they have long been here and we will not need much effort to support them (though I don't think they are useful). The deprecation is proposed by lesteve, you might ping him for more reasons. Or if you are pretty sure, please give a clear instruction and I'll remove the deprecation and leave this PR as a doc improvement :)

jnothman commented:
I can wait until @lesteve replies :)

lesteve commented:
Apologies for taking some time to come back to this one. 

In other words, a way to put this in plots:

```py
import matplotlib.pyplot as plt
plt.figure()
plt.plot([0, 1, 1], [0, 0, 1], 'o-')
plt.figure()
# this curve has the two last points swapped
plt.plot([0, 1, 1], [0, 1, 0], 'o-')
plt.show()
```

![figure_1](https://user-images.githubusercontent.com/1680079/31390427-25bbb3f0-add4-11e7-8eb6-72cb71c1feb4.png)

![figure_2](https://user-images.githubusercontent.com/1680079/31390436-2adc8a6c-add4-11e7-984d-b331c165a51f.png)

Before #9786 was merged we would be switching from the first curve to the second curve because of floating point noise in `x` and `reorder=True`. The areas under the curve of this two curves are very different and this was the reason we were seeing `roc_auc_score` differences a lot bigger than what you would expect from floating point differences.

In the case we fixed in https://github.com/scikit-learn/scikit-learn/pull/9786, fpr was not increasing only for floating point reasons. In fact we were guaranteed that fpr were in the right order because they were computed with decreasing thresholds. `reorder=True` was an easy way to get rid of the error "x is not increasing" but the wrong thing to do because rearranging the curve as `reorder=True` does was drastically changing the auc score (as the plots above are trying to illustrate). This is the main reason why I would be in favour of dropping the `reorder` parameter in the long run.

qinhanmin2014 commented:
@lesteve Thanks a lot for the detailed reply :)
So I think your opinion is that it is better for us to force users to pass increasing/decreasing x so that we can have a sanity check to avoid wrong result, right? Considering that the result from auc will be significantly influenced when x is sorted unexpectedly due to slight floating point error. I think at this point at least you persuade me to do the deprecation. I'll update the reason for deprecation.
ping @jnothman

qinhanmin2014 committed:
update reason for deprecation

qinhanmin2014 committed:
update reason for deprecation

jnothman commented:
I suppose you think users are as foolish as we are, to think that setting
reorder=True might be an answer to their problems. maybe then tweaking the
error message to say "this may be a result of numerical instability in
which case reorder=True is not a recommended solution"

qinhanmin2014 commented:
ping @lesteve 
I think I'll wait for the final decision from @jnothman and @lesteve.
Personally, after @lesteve's explanation, I might prefer to deprecate reorder. It seems worthwhile to force users to change their code since the potential error might be significant. Also, another important reason is that, I don't think the reorder parameter is designed for general case, because sometimes users might don't want to use y to break ties (like the case below). So it might be better that users sort x and arrange y themselves to get the result they expect.
![index](https://user-images.githubusercontent.com/12003569/31419773-40b1c770-ae70-11e7-841a-dd94172a7bb2.png)
Thanks for the the detailed instructions from all of you :)

jnothman commented:
"ascending in case of ties" really should be stated as "y has a monotonic
relation to x".

I really don't mind much here. Fine. Get rid of reorder.

On 11 October 2017 at 13:43, Hanmin Qin <notifications@github.com> wrote:

> ping @lesteve <https://github.com/lesteve>
> I think I'll wait for the final decision from @jnothman
> <https://github.com/jnothman> and @lesteve <https://github.com/lesteve>.
> Personally, after @lesteve <https://github.com/lesteve>'s explanation, I
> might prefer to deprecate reorder. It seems worthwhile to force users to
> change their code since the potential error might be significant. Also,
> another important reason is that, I don't think the reorder parameter is
> designed for general case, because sometimes users might don't want to use
> y to break ties (like the case below). So it might be better that users
> sort x and arrange y themselves to get the result they expect.
> [image: index]
> <https://user-images.githubusercontent.com/12003569/31419773-40b1c770-ae70-11e7-841a-dd94172a7bb2.png>
> Thanks for the the detailed instructions from all of you :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9851#issuecomment-335665465>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6wwvhpy5bI7UF1r84rSc7bWAlJJmks5srCs_gaJpZM4PoKta>
> .
>

qinhanmin2014 committed:
more deprecation reasons

qinhanmin2014 commented:
@jnothman I have updated a more detailed reason for deprecation. Do you think it is enough for you to remove the `?` in the title? Thanks :)

qinhanmin2014 committed:
more comment according to jnothman

jnothman commented:
Sure

lesteve commented on sklearn/metrics/ranking.py:
```diff
@@ -47,12 +47,23 @@ def auc(x, y, reorder=False):
+          general use) and is no longer used there. What's more, the result
+          from auc will be significantly influenced if x is sorted
+          unexpectedly due to slight floating point error (See issue #9786).
+          Future (and default) behavior is equivalent to `reorder=False`.
```
This is rst and not markdown so you need double-backticks around reorder=False. Single backtick is italic in rst.

lesteve commented on sklearn/metrics/ranking.py:
```diff
@@ -83,8 +94,13 @@ def auc(x, y, reorder=False):
 
+    if reorder != 'deprecated':
+        warnings.warn("The `reorder` parameter has been deprecated "
+                      "in version 0.20 and will be removed in 0.22.",
```
Best practices: the warning ideally:
* explains why there is a warning (maybe here the doc + release notes is enough in this case)
* what to do to get rid of it. In this case say it is recommended to not set reorder (same behaviour as reorder=False) and make sure to have x and y ordered (maybe add as the points of a curve. I am not sure how to phrase this better ...)

lesteve reviewed (commented):
Some comments, otherwise LGTM.

qinhanmin2014 committed:
more informative warning message

qinhanmin2014 commented:
@lesteve Thanks :) Comments addressed.

qinhanmin2014 committed:
fix stupid mistake

lesteve commented:
@jnothman any comments on this one? If I don't hear anything I'll merge it in a few days.

jnothman committed:
DOC attempt to clarify what's new wording

jnothman reviewed (commented):
LGTM

lesteve commented:
Good stuff @qinhanmin2014!
